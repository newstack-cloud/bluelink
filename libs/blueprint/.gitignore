# Code generated pre-build, (e.g genny generics templates)
gen-*.go

# Binaries for programs and plugins
*.exe
*.exe~
*.dll
*.so
*.dylib

# Test binary, built with `go test -c`
*.test

# Output of the go coverage tool, specifically when used with LiteIDE
*.out
coverage.txt
coverage.html

# Test report for sonar.
report.json

# Dependency directories (remove the comment below to include it)
# vendor/

# Any log files.
*.log

# Directory for programs for manually testing blueprint framework features.
playground

# Any temporary documentation files that shouldn't be in version control.
tmp-*.md
