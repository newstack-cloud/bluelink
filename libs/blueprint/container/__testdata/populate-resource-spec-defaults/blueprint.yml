version: 2025-05-12
variables:
  instanceType:
    type: aws/ec2/instanceType
    description: "The configuration buckets to create."
  environment:
    type: string

values:
  tableName:
    type: string
    value: "${variables.environment}-ordersTable"

datasources:
  network:
    type: aws/vpc
    description: "Networking resources for the application."
    filter:
      field: tags
      operator: "not contains"
      search: service
    metadata:
      displayName: Networking
    exports:
      vpc:
        type: string
        aliasFor: vpcId
        description: |
          The ID of the VPC.
      subnetIds:
        type: array
        description: "The IDs of the subnets."

resources:
  complexResource:
    type: example/complex
    description: "A complex resource type for which defaults should be populated."
    spec:
      name: "Orders"
      itemConfig:
        endpoints:
          - "https://example.com/1"
          - "https://example.com/2"
        ipv4: true
        # primaryPort should be populated with a default value.
        metadata:
          key1: value1
          key2: value2
      vendorConfig:
        - vendorNamespace: "example-vendor-namespace"
          # vendorId should be populated with a default value.
      # otherItemConfig should be populated with a default value.

include:
  coreInfra:
    path: core-infra.yaml
    description: core infrastructure for the Orders API
    metadata:
      sourceType: aws/s3
      bucket: order-system-blueprints
      region: eu-west-1

exports:
  environment:
    type: string
    field: variables.environment
    description: "The environment for the blueprint."

metadata:
  build: esbuild
