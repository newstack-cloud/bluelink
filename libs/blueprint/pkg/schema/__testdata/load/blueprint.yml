version: 2021-12-18
transform:
  - celerity-2022-01-22
variables:
  dynamoDBTable:
    type: string
    description: The Orders DynamoDB Table
  ordersTopicName:
    type: string
    description: The Orders Event Topic
dataSources:
  network:
    type: "aws/vpc"
    metadata:
      displayName: Network source
    exports:
      subnets:
        type: array
      securityGroups:
        type: array
      vpc:
        type: string
resources:
  # An API resource must be defined even if there are no endpoint handlers
  # for an application as env vars, runtime and tracing config for all handlers
  # are defined at the API level.
  orderApi:
    # For Makeshift to be fully compatible with celerity, we'll use the
    # celerity namespace for all Makeshift resources.
    type: "celerity/api"
    metadata:
      displayName: Order API
      labels:
        app: orderApi
      custom:
        visual:
          point: 20 100
          width: 50
          height: 50
          linkPorts:
            - linkTo: getOrderEndpoint
              connectAt: 70 100
              annotation: Orders API
    linkSelector:
      byLabel:
        app: orderApi
    spec:
      # Environment variables are set at the API level so the spec is compatible
      # with a containerised deployment without undesired, confusing behaviour
      # when you define env vars for individual handlers and some values would
      # be overridden when deployed as an API server.
      #
      # The linking behaviour also injects environment variables to be used at runtime,
      # these environment variables are displayed in the `deploy` and `dry-run` commands
      # and can be exported to a file, so you know what environment variables you can
      # reference from code.
      environment:
        variables:
          DYNAMODB_TABLE: ${dynamoDBTable}
      # All handlers need to use the same runtime to be compatible with a containerised
      # deployment.
      # The closest equivalent runtime will be selected for each cloud provider.
      # If a runtime is not supported by the given provider, an error will occur.
      runtime: go1.x
      # Tracing needs to be set API-wide to be compatible with a containerised
      # deployment. This will also enable tracing for handlers.
      tracingEnabled: true

  # In the serverless deployments this will be an AWS Lambda Authoriser, as the function
  # will be wrapped by a Makeshift/Celerity Handler wrapper it will be adapted into the Lambda interface
  # for authorisers.
  # In containerised API server deployments this will be a middleware, as handlers and middleware
  # have exactly the same interface, this will work perfectly.
  authoriser:
    type: "celerity/handler"
    metadata:
      displayName: Authoriser
      annotations:
        # Gives more context for the API resource
        # that links to this handler so it knows to use
        # it as an authoriser and not an endpoint.
        authoriser: true
        # Custom annotations that enrich the handler with authorisation
        # configuration that will be picked up to configure
        # the authoriser when deploying the API.
        authoriser.identity.header: MyCustomAuthHeader
        authoriser.identity.validationExpression: "Bearer .*"
        authoriser.identity.reauthoriseEvery: 20
      labels:
        app: orderApi
    spec:
      handler: handlers.Authoriser
      timeout: 120

  getOrdersHandler:
    type: "celerity/handler"
    metadata:
      displayName: Get Orders Handler
      labels:
        app: orderApi
    spec:
      handler: handlers.GetOrders
      timeout: 120
      endpoint:
        method: get
        path: /orders
        authoriser: ${authoriser}

  # This is a consumer handler function.
  updateOrderFunction:
    type: "celerity/handler"
    metadata:
      displayName: Update Order Function
      labels:
        workflow: orderPubSub
        # This is required so the linking logic knows the type of relationship
        # due to the fact that this handler can have it's own linkSelector that
        # can select by the same label and handlers can have a publisher and
        # consumer relationship with a pubsub resource.
        pubsubType: consumer
    spec:
      handler: handlers.UpdateOrder
      timeout: 120
      events:
        orderEvent:
          type: pubsub
          # The pub/sub topic is automatically linked with the
          # resource via the link selector.
          batchSize: 10

  sharedPackagesLayer:
    # When the provider is AWS and the deployment target is serverless,
    # this becomes a lambda layer.
    # Layers are bundled in the root directory of a serverless function or
    # a container in every other situation. For containers, the makeshift
    # CLI will augment the Dockerfile for your project to make sure these
    # packages are included if they aren't already.
    type: "celerity/layerVersion"
    metadata:
      displayName: Shared Packages Layer
      labels:
        workflow: orderPubSub
        app: orderApi
    spec:
      compatibleRuntimes: ["go1.x"]
      # A file path on the local file system where the spec
      # is being deployed.
      content: build/shared-packages-layer
      layerName: sharedPackagesLayer
  orderPubSub:
    # This creates an SNS topic (if it doesn't already exist) and an SQS queue
    # subscription in AWS. For compatibility with Google Cloud pub/sub which will
    # create a single resource we use the notion of pub/sub for the entire resource.
    type: "celerity/pubsub"
    metadata:
      displayName: Order Pub/Sub
      labels:
        workflow: orderPubSub
    linkSelector:
      byLabel:
        workflow: orderPubSub
    spec:
      topicName: ${ordersTopicName}
