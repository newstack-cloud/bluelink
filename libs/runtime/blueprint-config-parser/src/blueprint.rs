use std::collections::HashMap;

use serde::{Deserialize, Serialize};

/// The default version for the Celerity blueprint configuration.
pub const CELERITY_BLUEPRINT_V2023_04_20: &str = "2023-04-20";

/// The resource type identifier for a Celerity API.
pub const CELERITY_API_RESOURCE_TYPE: &str = "celerity/api";

/// The resource type identifier for a Celerity Consumer.
pub const CELERITY_CONSUMER_RESOURCE_TYPE: &str = "celerity/consumer";

/// The resource type identifier for a Celerity Schedule.
pub const CELERITY_SCHEDULE_RESOURCE_TYPE: &str = "celerity/schedule";

/// The resource type identifier for a Celerity Handler.
pub const CELERITY_HANDLER_RESOURCE_TYPE: &str = "celerity/handler";

/// The resource type identifier a Celerity Handler Config (shared config).
pub const CELERITY_HANDLER_CONFIG_RESOURCE_TYPE: &str = "celerity/handlerConfig";

/// This is a struct that holds the configuration
/// for the Celerity runtime in the form a blueprint.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct BlueprintConfig {
    #[serde(deserialize_with = "crate::parse_helpers::deserialize_version")]
    pub version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(deserialize_with = "crate::parse_helpers::deserialize_optional_string_or_vec")]
    pub transform: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variables: Option<HashMap<String, BlueprintVariable>>,
    #[serde(deserialize_with = "crate::parse_helpers::deserialize_resource_map")]
    pub resources: HashMap<String, RuntimeBlueprintResource>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<BlueprintMetadata>,
}

impl Default for BlueprintConfig {
    fn default() -> Self {
        BlueprintConfig {
            version: "".to_string(),
            transform: None,
            variables: None,
            resources: HashMap::new(),
            metadata: None,
        }
    }
}

/// This is a struct that holds a variable
/// in the blueprint configuration.
/// In the runtime, variables can be sourced
/// from the runtime environment and used in the evaluation
/// of the blueprint configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct BlueprintVariable {
    #[serde(rename = "type")]
    pub var_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "allowedValues")]
    pub allowed_values: Option<Vec<BlueprintScalarValue>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default: Option<BlueprintScalarValue>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub secret: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

impl Default for BlueprintVariable {
    fn default() -> Self {
        BlueprintVariable {
            var_type: "".to_string(),
            allowed_values: None,
            default: None,
            secret: None,
            description: None,
        }
    }
}

/// This is a struct that holds a value
/// for a scalar that can be multiple types
/// in the blueprint configuration.
/// This is used to define allowed and default values
/// for a variable.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
#[serde(untagged)]
pub enum BlueprintScalarValue {
    Str(String),
    Int(i64),
    Float(f64),
    Bool(bool),
}

impl ToString for BlueprintScalarValue {
    fn to_string(&self) -> String {
        match self {
            BlueprintScalarValue::Str(val) => val.to_string(),
            BlueprintScalarValue::Int(val) => val.to_string(),
            BlueprintScalarValue::Float(val) => val.to_string(),
            BlueprintScalarValue::Bool(val) => val.to_string(),
        }
    }
}

/// This is a struct that holds the configuration
/// for a resource in the blueprint configuration.
/// This is a type specific to the runtime configuration
/// that is focused on the Celerity-specific resources
/// that are used in the blueprint configuration.
/// Resource types that are not recognised by the runtime
/// will be ignored.
#[derive(Serialize, Debug, PartialEq)]
pub struct RuntimeBlueprintResource {
    #[serde(rename = "type")]
    pub resource_type: CelerityResourceType,
    pub metadata: BlueprintResourceMetadata,
    pub spec: CelerityResourceSpec,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "linkSelector")]
    pub link_selector: Option<BlueprintLinkSelector>,
}

impl Default for RuntimeBlueprintResource {
    fn default() -> Self {
        RuntimeBlueprintResource {
            resource_type: CelerityResourceType::CelerityHandler,
            metadata: BlueprintResourceMetadata {
                display_name: "".to_string(),
                annotations: None,
                labels: None,
            },
            link_selector: None,
            description: None,
            spec: CelerityResourceSpec::NoSpec,
        }
    }
}

/// This is an enum that holds the types of resources
/// that are recognised by the Celerity runtime.
/// The runtime will only process resources that are
/// of these types and ignore any other types.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum CelerityResourceType {
    #[serde(rename = "celerity/handler")]
    CelerityHandler,
    #[serde(rename = "celerity/api")]
    CelerityApi,
    #[serde(rename = "celerity/consumer")]
    CelerityConsumer,
    #[serde(rename = "celerity/schedule")]
    CeleritySchedule,
    #[serde(rename = "celerity/handlerConfig")]
    CelerityHandlerConfig,
}

/// This holds the metadata
/// for a resource in the blueprint configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct BlueprintResourceMetadata {
    #[serde(rename = "displayName")]
    pub display_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub annotations: Option<HashMap<String, BlueprintScalarValue>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub labels: Option<HashMap<String, String>>,
}

impl Default for BlueprintResourceMetadata {
    fn default() -> Self {
        BlueprintResourceMetadata {
            display_name: "".to_string(),
            annotations: None,
            labels: None,
        }
    }
}

/// This holds the configuration
/// for a link selector in the blueprint configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq, Default)]
pub struct BlueprintLinkSelector {
    #[serde(rename = "byLabel")]
    pub by_label: HashMap<String, String>,
}

/// This holds the specification
/// for a resource in the blueprint configuration.
/// This is specific to resource types recognised
/// by the Celerity runtime.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
#[serde(untagged)]
pub enum CelerityResourceSpec {
    Handler(CelerityHandlerSpec),
    Api(CelerityApiSpec),
    Consumer(CelerityConsumerSpec),
    Schedule(CelerityScheduleSpec),
    HandlerConfig(SharedHandlerConfig),
    NoSpec,
}

/// This holds the specification
/// for a handler resource in the blueprint configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct CelerityHandlerSpec {
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "handlerName")]
    pub handler_name: Option<String>,
    #[serde(rename = "codeLocation")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code_location: Option<String>,
    pub handler: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub runtime: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub memory: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "tracingEnabled")]
    pub tracing_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "environmentVariables")]
    pub environment_variables: Option<HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub events: Option<HashMap<String, EventConfiguration>>,
}

impl Default for CelerityHandlerSpec {
    fn default() -> Self {
        CelerityHandlerSpec {
            handler_name: None,
            code_location: None,
            handler: "".to_string(),
            runtime: None,
            memory: None,
            timeout: None,
            tracing_enabled: None,
            environment_variables: None,
            events: None,
        }
    }
}

/// This holds the specification
/// for an API resource in the blueprint configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq, Default)]
pub struct CelerityApiSpec {
    pub protocols: Vec<CelerityApiProtocol>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cors: Option<CelerityApiCors>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain: Option<CelerityApiDomain>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auth: Option<CelerityApiAuth>,
    #[serde(rename = "tracingEnabled")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tracing_enabled: Option<bool>,
}

/// This holds the specification
/// for a consumer resource in the blueprint configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct CelerityConsumerSpec {
    #[serde(rename = "sourceId")]
    pub source_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "batchSize")]
    pub batch_size: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "visibilityTimeout")]
    pub visibility_timeout: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "waitTimeSeconds")]
    pub wait_time_seconds: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "partialFailures")]
    pub partial_failures: Option<bool>,
}

impl Default for CelerityConsumerSpec {
    fn default() -> Self {
        CelerityConsumerSpec {
            source_id: "".to_string(),
            batch_size: None,
            visibility_timeout: None,
            wait_time_seconds: None,
            partial_failures: None,
        }
    }
}

/// This holds the specification
/// for a schedule resource in the blueprint configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct CelerityScheduleSpec {
    pub schedule: String,
}

impl Default for CelerityScheduleSpec {
    fn default() -> Self {
        CelerityScheduleSpec {
            schedule: "".to_string(),
        }
    }
}

/// A protocol that an API resource can support.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum CelerityApiProtocol {
    #[serde(rename = "http")]
    Http,
    #[serde(rename = "websocket")]
    WebSocket,
}

/// CORS configuration for a Celerity API resource which can be
/// a string or detailed configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
#[serde(untagged)]
pub enum CelerityApiCors {
    Str(String),
    CorsConfiguration(CelerityApiCorsConfiguration),
}

/// Detailed CORS configuration
/// for a Celerity API resource.
#[derive(Serialize, Deserialize, Debug, PartialEq, Default)]
pub struct CelerityApiCorsConfiguration {
    #[serde(rename = "allowCredentials")]
    pub allow_credentials: Option<bool>,
    #[serde(rename = "allowOrigins")]
    pub allow_origins: Option<Vec<String>>,
    #[serde(rename = "allowMethods")]
    pub allow_methods: Option<Vec<String>>,
    #[serde(rename = "allowHeaders")]
    pub allow_headers: Option<Vec<String>>,
    #[serde(rename = "exposeHeaders")]
    pub expose_headers: Option<Vec<String>>,
    #[serde(rename = "maxAge")]
    pub max_age: Option<i64>,
}

/// Domain configuration for a Celerity API resource.
#[derive(Serialize, Deserialize, Debug, PartialEq, Default)]
pub struct CelerityApiDomain {
    #[serde(rename = "domainName")]
    pub domain_name: String,
    #[serde(rename = "basePaths")]
    pub base_paths: Vec<CelerityApiBasePath>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "normalizeBasePath")]
    pub normalize_base_path: Option<bool>,
    #[serde(rename = "certificateId")]
    pub certificate_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "securityPolicy")]
    pub security_policy: Option<CelerityApiDomainSecurityPolicy>,
}

/// Base path configuration for a Celerity API resource which can be
/// a string or detailed configuration.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
#[serde(untagged)]
pub enum CelerityApiBasePath {
    Str(String),
    BasePathConfiguration(CelerityApiBasePathConfiguration),
}

/// Base path configuration for a Celerity API resource.
/// This allows you to configure a base path for a specific
/// protocol for an API.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct CelerityApiBasePathConfiguration {
    pub protocol: CelerityApiProtocol,
    #[serde(rename = "basePath")]
    pub base_path: String,
}

impl Default for CelerityApiBasePathConfiguration {
    fn default() -> Self {
        CelerityApiBasePathConfiguration {
            protocol: CelerityApiProtocol::Http,
            base_path: "".to_string(),
        }
    }
}

/// Security policy for a Celerity API domain.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum CelerityApiDomainSecurityPolicy {
    #[serde(rename = "TLS_1_0")]
    Tls1_0,
    #[serde(rename = "TLS_1_2")]
    Tls1_2,
}

/// Authentication configuration for a Celerity API resource.
#[derive(Serialize, Deserialize, Debug, PartialEq, Default)]
pub struct CelerityApiAuth {
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "defaultGuard")]
    pub default_guard: Option<String>,
    pub guards: HashMap<String, CelerityApiAuthGuard>,
}

/// Guard configuration that provides access control
/// for a Celerity API resource.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct CelerityApiAuthGuard {
    #[serde(rename = "type")]
    pub guard_type: CelerityApiAuthGuardType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub issuer: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "tokenSource")]
    pub token_source: Option<CelerityApiAuthGuardValueSource>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audience: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "apiKeySource")]
    pub api_key_source: Option<CelerityApiAuthGuardValueSource>,
}

impl Default for CelerityApiAuthGuard {
    fn default() -> Self {
        CelerityApiAuthGuard {
            guard_type: CelerityApiAuthGuardType::NoGuardType,
            issuer: None,
            token_source: None,
            audience: None,
            api_key_source: None,
        }
    }
}

/// Auth guard type for authorization configuration
/// in a Celerity API resource.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum CelerityApiAuthGuardType {
    #[serde(rename = "jwt")]
    Jwt,
    #[serde(rename = "apiKey")]
    ApiKey,
    #[serde(rename = "custom")]
    Custom,
    #[serde(rename = "noGuardType")]
    NoGuardType,
}

/// Value source for authorization configuration
/// in a Celerity API resource.
/// A value would be an API key or JWT.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
#[serde(untagged)]
pub enum CelerityApiAuthGuardValueSource {
    Str(String),
    ValueSourceConfiguration(ValueSourceConfiguration),
}

/// Value source configuration for extracting a value
/// from a request or message.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct ValueSourceConfiguration {
    pub protocol: CelerityApiProtocol,
    pub source: String,
}

impl Default for ValueSourceConfiguration {
    fn default() -> Self {
        ValueSourceConfiguration {
            protocol: CelerityApiProtocol::Http,
            source: "".to_string(),
        }
    }
}

/// Configuration for a cloud service event source
/// for a handler resource.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct EventConfiguration {
    #[serde(rename = "sourceType")]
    pub source_type: EventSourceType,
    #[serde(rename = "sourceConfiguration")]
    pub source_configuration: EventSourceConfiguration,
}

impl Default for EventConfiguration {
    fn default() -> Self {
        EventConfiguration {
            source_type: EventSourceType::ObjectStorage,
            source_configuration: EventSourceConfiguration::ObjectStorage(
                ObjectStorageEventSourceConfiguration {
                    bucket: "".to_string(),
                    events: vec![],
                },
            ),
        }
    }
}

/// The type of event source for a handler resource.
/// This can be an object storage, database stream
/// or a data stream.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum EventSourceType {
    #[serde(rename = "objectStorage")]
    ObjectStorage,
    #[serde(rename = "dbStream")]
    DatabaseStream,
    #[serde(rename = "dataStream")]
    DataStream,
}

/// Configuration for an event source for a handler resource.
/// This can be a configuration for an object storage,
/// database stream or a data stream.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
#[serde(untagged)]
pub enum EventSourceConfiguration {
    ObjectStorage(ObjectStorageEventSourceConfiguration),
    DatabaseStream(DatabaseStreamSourceConfiguration),
    DataStream(DataStreamSourceConfiguration),
}

/// Configuration for an object storage event source
/// for a handler resource.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct ObjectStorageEventSourceConfiguration {
    pub bucket: String,
    pub events: Vec<ObjectStorageEventType>,
}

impl Default for ObjectStorageEventSourceConfiguration {
    fn default() -> Self {
        ObjectStorageEventSourceConfiguration {
            bucket: "".to_string(),
            events: vec![],
        }
    }
}

/// Event types for an object storage event source.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum ObjectStorageEventType {
    #[serde(rename = "created")]
    ObjectCreated,
    #[serde(rename = "deleted")]
    ObjectDeleted,
    #[serde(rename = "metadataUpdated")]
    ObjectMetadataUpdated,
}

/// Configuration for a database stream event source
/// for a handler resource.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct DatabaseStreamSourceConfiguration {
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "batchSize")]
    pub batch_size: Option<i64>,
    #[serde(rename = "dbStreamId")]
    pub db_stream_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "partialFailures")]
    pub partial_failures: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "startFromBeginning")]
    pub start_from_beginning: Option<bool>,
}

impl Default for DatabaseStreamSourceConfiguration {
    fn default() -> Self {
        DatabaseStreamSourceConfiguration {
            batch_size: None,
            db_stream_id: "".to_string(),
            partial_failures: None,
            start_from_beginning: None,
        }
    }
}

/// Configuration for a data stream event source
/// for a handler resource.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct DataStreamSourceConfiguration {
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "batchSize")]
    pub batch_size: Option<i64>,
    #[serde(rename = "dataStreamId")]
    pub data_stream_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "partialFailures")]
    pub partial_failures: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "startFromBeginning")]
    pub start_from_beginning: Option<bool>,
}

impl Default for DataStreamSourceConfiguration {
    fn default() -> Self {
        DataStreamSourceConfiguration {
            batch_size: None,
            data_stream_id: "".to_string(),
            partial_failures: None,
            start_from_beginning: None,
        }
    }
}

/// Metadata for a blueprint.
/// For the purpose of the runtime, this is strongly
/// typed to expect an optional `sharedHandlerConfig`
/// object that provides shared defaults for all handlers
/// declared in a blueprint.
#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub struct BlueprintMetadata {
    #[serde(rename = "sharedHandlerConfig")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub shared_handler_config: Option<SharedHandlerConfig>,
}

/// Provides shared defaults
/// for all handlers declared in a blueprint.
#[derive(Serialize, Deserialize, Debug, PartialEq, Default)]
pub struct SharedHandlerConfig {
    #[serde(rename = "codeLocation")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code_location: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub runtime: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub memory: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "tracingEnabled")]
    pub tracing_enabled: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "environmentVariables")]
    pub environment_variables: Option<HashMap<String, String>>,
}
