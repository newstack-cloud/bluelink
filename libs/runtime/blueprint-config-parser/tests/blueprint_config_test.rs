use insta::{assert_json_snapshot, with_settings};
use std::fs::read_to_string;

use celerity_blueprint_config_parser::blueprint::BlueprintConfig;

#[test_log::test]
fn parses_blueprint_config_from_yaml_string() {
    let doc_str: String = read_to_string("tests/data/fixtures/http-api.yaml").unwrap();
    let blueprint_config = BlueprintConfig::from_yaml_str(doc_str.as_str()).unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}

#[test_log::test]
fn parses_blueprint_config_from_json_string() {
    let doc_str: String = read_to_string("tests/data/fixtures/http-api.json").unwrap();
    let blueprint_config = BlueprintConfig::from_json_str(doc_str.as_str()).unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}

#[test_log::test]
fn parses_http_api_blueprint_config_from_yaml_file() {
    let blueprint_config =
        BlueprintConfig::from_yaml_file("tests/data/fixtures/http-api.yaml").unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}

#[test_log::test]
fn parses_http_api_blueprint_config_from_json_file() {
    let blueprint_config =
        BlueprintConfig::from_json_file("tests/data/fixtures/http-api.json").unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}

#[test_log::test]
fn parses_websocket_api_blueprint_config_from_yaml_file() {
    let blueprint_config =
        BlueprintConfig::from_yaml_file("tests/data/fixtures/websocket-api.yaml").unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}

#[test_log::test]
fn parses_websocket_api_blueprint_config_from_json_file() {
    let blueprint_config =
        BlueprintConfig::from_json_file("tests/data/fixtures/websocket-api.json").unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}

#[test_log::test]
fn parses_combined_app_blueprint_config_from_yaml_file() {
    let blueprint_config =
        BlueprintConfig::from_yaml_file("tests/data/fixtures/combined-app.yaml").unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}

#[test_log::test]
fn parses_combined_app_blueprint_config_from_json_file() {
    let blueprint_config =
        BlueprintConfig::from_json_file("tests/data/fixtures/combined-app.json").unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}

#[test_log::test]
fn parses_hybrid_api_blueprint_config_from_yaml_file() {
    let blueprint_config =
        BlueprintConfig::from_yaml_file("tests/data/fixtures/hybrid-api.yaml").unwrap();

    with_settings!({sort_maps => true}, {
        assert_json_snapshot!(blueprint_config);
    })
}
