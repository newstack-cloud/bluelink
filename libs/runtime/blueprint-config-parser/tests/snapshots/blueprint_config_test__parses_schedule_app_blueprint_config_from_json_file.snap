---
source: blueprint-config-parser/tests/blueprint_config_test.rs
expression: blueprint_config
---
{
  "version": "2023-04-20",
  "transform": [
    "celerity-2024-07-22"
  ],
  "resources": {
    "jobHandler": {
      "type": "celerity/handler",
      "metadata": {
        "displayName": "Job Handler",
        "annotations": {
          "celerity.handler.schedule": true
        },
        "labels": {
          "application": "jobs"
        }
      },
      "spec": {
        "handlerName": "Jobs-JobHandler-v1",
        "codeLocation": "./jobs",
        "handler": "handlers.job",
        "runtime": "python3.12.x",
        "memory": 1024,
        "timeout": 60,
        "tracingEnabled": true,
        "environmentVariables": {
          "LOG_LEVEL": "INFO"
        }
      }
    },
    "jobSchedule": {
      "type": "celerity/schedule",
      "metadata": {
        "displayName": "Job Schedule"
      },
      "spec": {
        "schedule": "rate(1 hour)"
      },
      "linkSelector": {
        "byLabel": {
          "application": "jobs"
        }
      }
    }
  }
}
